packages/csvviewer/src/model.ts

```diff
@@ -0,0 +1,663 @@
+    this._delayedParse = window.setTimeout(delayedParse, delay);
+  }
+
+  /**
+   * Reset the parser state.
+   */
+  private _resetParser() {
+    this._columnCount = undefined;
+
+    // First row offset is *always* 0, so we always have the first row offset.
+    this._rowOffsets = new Uint32Array(1);
+    this._rowCount = 1;
+
+    this._columnOffsets = new Uint32Array(0);
+
+    // Clear out state associated with the asynchronous parsing.
+    if (this._doneParsing === false) {
+      // Explicitly catch this rejection at least once so an error is not thrown
+      // to the console.
+      this.ready.catch(() => { return; });
+      this._ready.reject(undefined);
+    }
+    this._doneParsing = false;
+    this._ready = new PromiseDelegate<void>();
+    if (this._delayedParse !== null) {
+      window.clearTimeout(this._delayedParse);
+      this._delayedParse = null;
+    }
+
+    this.emitChanged({ type: 'model-reset' });
+  }
+
+  // Parser settings
+  private _delimiter: string;
+  private _quote: string;
+  private _quoteEscaped: RegExp;
+  private _parser: 'quotes' | 'noquotes';
+  private _rowDelimiter: string;
+
+  // Data values
+  private _data: string;
+  private _rowCount: number = 1;
+  private _columnCount: number;
+
+  // Cache information
+  /**
+   * The header strings.
+   */
+  private _header: string[] = [];
+  /**
+   * The column offset cache, starting with row _columnOffsetsStartingRow
+   *
+   * #### Notes
+   * The index of the first character in the data string for row r, column c is
+   * _columnOffsets[(r-this._columnOffsetsStartingRow)*numColumns+c]
+   */
+  private _columnOffsets: Uint32Array = new Uint32Array(0);
+  /**
+   * The row that _columnOffsets[0] represents.
+   */
+  private _columnOffsetsStartingRow: number = 0;
+  /**
+   * The maximum number of rows to parse when there is a cache miss.
+   */
+  private _maxCacheGet: number = 1000;
+  /**
+   * The index for the start of each row.
+   */
+  private _rowOffsets: Uint32Array = new Uint32Array(1);
+  /**
+   * The number of rows to parse initially before doing a delayed parse of the
+   * entire data.
+   */
+  private _initialRows: number;
+
+
+  // Bookkeeping variables.
+  private _delayedParse: number = null;
+  private _doneParsing: boolean = false;
+  private _isDisposed: boolean = false;
+  private _ready = new PromiseDelegate<void>();
+}
+
+
+/**
+ * The namespace for the `DSVModel` class statics.
+ */
+export
+namespace DSVModel {
+
+  /**
+   * An options object for initializing a delimiter-separated data model.
+   */
+  export
+  interface IOptions {
+    /**
+     * The field delimiter, such as ',' or '\t'.
+     *
+     * #### Notes
+     * The field delimiter must be a single character.
+     */
+    delimiter: string;
+
+    /**
+     * The data source for the data model.
+     *
+     * The data model takes full ownership of the data source.
```

Comment:
Strings are immutable, so "ownership" doesn't really matter.

Reply:
You're right, that doesn't make sense. I don't know what I was thinking there.